test/test_scrypt.sh

```diff
@@ -30,79 +33,192 @@ if [ ! -f $scrypt_binary ] || [ ! -f $test_scrypt_binary ] || \
 	exit 1
 fi
 
-# Test functions
+if [ -z "$USE_VALGRIND" ]; then
+	USE_VALGRIND="0"
+fi
+if [ "$USE_VALGRIND" -gt 0 ]; then
+	if ! command -v valgrind >/dev/null 2>&1; then
+		echo "valgrind not detected: disabling memory checking"
+		USE_VALGRIND="0"
+	fi
+fi
+
+################################ Helper functions
+
+setup_valgrind_cmd() {
+	basename=$1
+	# The user-specified $USE_VALGRIND number must be higher than
+	# $valgrind_disable_number; this allows us to specify certain tests as
+	# being "normal memory usage" or "lots of memory required; most people
+	# won't want to run valgrind on this".
+	valgrind_disable_number=$2
+
+	# Set up valgrind command (if requested).  Using --error-exitcode
+	# means that if there is a serious problem (such that scrypt calls
+	# exit(1)) *and* a memory leak, the test suite reports an exit value
+	# of $valgrind_exit.  However, if there is a serious problem but no
+	# memory leak, we still receive a non-zero exit code.  The most
+	# important thing is that we only receive an exit code of 0 if both
+	# the program and valgrind are happy.
+	if [ "$USE_VALGRIND" -gt "$valgrind_disable_number" ]; then
+		valgrind_cmd="valgrind \
+			--log-file=$out_valgrind/$basename-val.log \
+			--leak-check=full --show-leak-kinds=all \
+			--errors-for-leak-kinds=all \
+			--error-exitcode=$valgrind_exit "
+	else
+		valgrind_cmd=""
+	fi
+
+	# Return command to calling function.
+	echo "$valgrind_cmd"
+}
+
+notify_success_or_fail() {
+	retval=$1
+	cmd_retval=$2
+	basename=$3
+
+	if [ "$retval" -eq 0 ]; then
+		echo "PASSED!"
+	else
+		echo "FAILED"
+		# If valgrind discovered a problem, print it.  This is
+		# primarily aimed at automation in travis-CI, since we don't
+		# have access to the log files.
+		if [ "$USE_VALGRIND" -gt 0 ]; then
```

Comment:
structuring it this way (i.e. without an `&&`) makes more sense when there's multiple `cmd_retval` and `basename` variables to check, as will be the case in spiped.  For scrypt, it's a bit of a wash, but I figured that reducing the difference between cperciva projects wouldn't be a bad thing.

... and yes, now I find myself wanting to have "posix sh libraries" so that I can re-use these functions.